### PR TITLE
Fix queries with FINAL give wrong result when table does not use adaptive granularity

### LINKED ISSUE #62426: final stopped working after 23.12 for ReplacingMergeTree  tables with a lot of partitions and records
Getting duplicate records for ReplacingMergeTree for queries with final modifier. 
Setting compatibility = '23.12' restores proper behavior.

Test case for 24.3 : https://fiddle.clickhouse.com/399466fd-51fa-45a1-a183-c13b5a0e41d5

Seems like it needs a lot of partitions and records.

### PR DESCRIPTION
**Pull request by canhld94**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Fix queries with FINAL give wrong result when table does not use adaptive granularity

Close #62426

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

---
### Modify your CI run:
**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step

#### Include tests (required builds will be added automatically):
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Unit tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with Analyzer
- [ ]  Add your option here

#### Exclude tests:
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with MSAN
- [ ]  All with UBSAN
- [ ]  All with Coverage
- [ ]  All with Aarch64
- [ ]  Add your option here

#### Extra options:
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ]  1
- [ ]  2
- [ ]  3
- [ ]  4

### CODE REVIEW

File: tests/queries/0_stateless/03033_final_undefined_last_mark.sql [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+         min_rows_for_wide_part = 0, min_bytes_for_wide_part = 0,
+         min_rows_for_compact_part = 0, min_bytes_for_compact_part = 0;
+
+INSERT INTO account_test
+	SELECT * FROM generateRandom('id UInt64, row_ver UInt64',1234) LIMIT 50000;
+
+INSERT INTO account_test
+    SELECT * FROM (SELECT * FROM generateRandom('id UInt64, row_ver UInt64',1234) LIMIT 1000) WHERE row_ver > 14098131981223776000;
+
+SELECT 'GOOD', * FROM account_test FINAL WHERE id = 11338881281426660955 SETTINGS split_parts_ranges_into_intersecting_and_non_intersecting_final = 1;
```
Comment (kitaisreal) [inline tests/queries/0_stateless/03033_final_undefined_last_mark.sql:21]: It is better to add another query in test:
```
SELECT 'GOOD', * FROM account_test FINAL WHERE id = 11338881281426660955 SETTINGS split_parts_ranges_into_intersecting_and_non_intersecting_final = 0;
```
So it is easier to compare outputs.

**Review by kitaisreal — CHANGES_REQUESTED** [review]

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit e8f02af78c418f7c0a521bd48d49fcfb91db455f with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/62432/e8f02af78c418f7c0a521bd48d49fcfb91db455f/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Review by kitaisreal — APPROVED** [review]

**Comment by canhld94** [comment]
AST Fuzzer test is failing, but I think it doesn't relates to the change this PR. Can reproduce in upstream `HEAD` even when turning off splitting parts https://fiddle.clickhouse.com/d5c476f8-9fb9-4ee1-bf10-0491e173bafb

**Comment by canhld94** [comment]
> AST Fuzzer test is failing, but I think it doesn't relates to the change this PR. Can reproduce in upstream `HEAD` even when turning off splitting parts https://fiddle.clickhouse.com/d5c476f8-9fb9-4ee1-bf10-0491e173bafb
> 
> Update: even without `FINAL`

#62741

**Comment by canhld94** [comment]
Stress test failure relates to  database replicated #62742

**Comment by canhld94** [comment]
Ping @Algunenano I think this PR is ready to merge

**Comment by alexey-milovidov** [comment]
@canhld94, what is with these logical errors detected by our fuzzers?

**Comment by canhld94** [comment]
> @canhld94, what is with these logical errors detected by our fuzzers?

@alexey-milovidov this is a known bug, I opened an issue for it https://github.com/ClickHouse/ClickHouse/pull/62432 and have an attempt to fix in https://github.com/ClickHouse/ClickHouse/pull/62776 but not yet found  a good approach.

**Comment by Algunenano** [comment]
Sorry for the delay reviewing this. Initially it slipped through the cracks and then I got sidetracked by bugs in related code (but not in this PR). I've simplified the test so it doesn't depend on generateRandom and updated with master to make sure things are still ok.

**Comment by canhld94** [comment]
> Sorry for the delay reviewing this. Initially it slipped through the cracks and then I got sidetracked by bugs in related code (but not in this PR). I've simplified the test so it doesn't depend on generateRandom and updated with master to make sure things are still ok.

Thanks! Nevertheless, the AST fuzzer failure is still there, not causing by this PR but merging this PR will reproduce the failure in master. May be we need to push to fix this bug soon.

**Pull request merged**